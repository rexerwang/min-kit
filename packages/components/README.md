# @min-kit/components

[![](https://img.shields.io/codecov/c/github/rexerwang/min-kit?flag=components&style=for-the-badge)](https://codecov.io/gh/rexerwang/min-kit/flags)
[![](https://img.shields.io/npm/types/%40min-kit/components?style=for-the-badge)](https://github.com/rexerwang/min-kit/tree/main/packages/components)
[![](https://img.shields.io/npm/v/%40min-kit/components?style=for-the-badge)](https://npm.im/@min-kit/components)
[![](https://img.shields.io/badge/React-Tarojs-007ACC?style=for-the-badge&logo=react&logoColor=61DAFB&labelColor=20232A)](https://github.com/NervJS/taro)

React components for miniapp

## Usage

```tsx
import { MinIcon, MinImage, MinNavigationBar, Modal, type ReactProps } from '@min-kit/components'
import { go } from '@min-kit/extends'
import { Button, View } from '@tarojs/components'

// pages/index/index.tsx
export default function Index(props: ReactProps) {
  const openModal = () => {
    Modal.confirm({
      title: 'Confirm',
      subtitle: 'the subtitle\nsupport multiple lines.',
      okText: 'ok',
      cancelText: 'cancel',
    })
  }

  return (
    <View className='relative box-border'>
      <MinNavigationBar
        title={
          <View className='flex-center-y'>
            <MinIcon name='back' size={20} onClick={() => go.back()} />
            <View className='ml-3 text-sm'>@min-kit/components</View>
          </View>
        }></MinNavigationBar>

      <View className='flex flex-col items-center'>
        <MinImage
          className='w-[750px] h-[750px]'
          src='https://source.unsplash.com/random/400x400?nature'
          preview={[
            'https://source.unsplash.com/random/400x400?nature',
            'https://source.unsplash.com/random/400x400?people',
            'https://source.unsplash.com/random/400x400?foods',
          ]}
        />
        <Button className='mt-4' type='primary' size='mini' plain onClick={openModal}>
          open modal
        </Button>
      </View>
    </View>
  )
}

definePageConfig({
  navigationStyle: 'custom',
})
```

### Import styles

```js
// Import overall styles
import '@min-kit/components/dist/styles/index.css'

// Import individual component styles
import '@min-kit/components/dist/styles/min-debugger.css'
import '@min-kit/components/dist/styles/min-icon.css'
import '@min-kit/components/dist/styles/min-image.css'
import '@min-kit/components/dist/styles/min-navigation-bar.css'
import '@min-kit/components/dist/styles/modal.css'

// Import SCSS
import '@min-kit/components/styles/index.scss'
import '@min-kit/components/styles/min-debugger.scss'
import '@min-kit/components/styles/min-icon.scss'
import '@min-kit/components/styles/min-image.scss'
import '@min-kit/components/styles/min-navigation-bar.scss'
import '@min-kit/components/styles/modal.scss'
```

More usage examples ðŸ‘‰ https://github.com/rexerwang/min-kit/tree/main/packages/example
