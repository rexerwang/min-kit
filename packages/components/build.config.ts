import { exec } from 'node:child_process'
import { resolve } from 'node:path'
import { promisify } from 'node:util'

import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  clean: !process.argv.includes('--no-clean'),
  entries: [
    'src/index',
    {
      builder: 'mkdist',
      input: 'styles',
      outDir: 'dist/styles',
    },
  ],
  declaration: true,
  rollup: {
    emitCJS: true,
    esbuild: {
      jsx: 'automatic',
    },
  },
  hooks: {
    async 'mkdist:done'() {
      const input = resolve(__dirname, 'dist/styles')
      await promisify(exec)(`pnpm postcss ${input} -d ${input}`)
    },
  },
})
