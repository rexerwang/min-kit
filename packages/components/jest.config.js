module.exports = require('@min-kit/jest/config')({
  collectCoverageFrom: ['!src/debugger/{hooks,service,store}/**'],
})
