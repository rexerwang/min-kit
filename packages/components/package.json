{
  "name": "@min-kit/components",
  "version": "0.5.1",
  "description": "React components for miniapp",
  "author": "rexerwang",
  "license": "MIT",
  "repository": "rexerwang/min-kit",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    },
    "./dist/styles/*.css": {
      "import": "./dist/styles/*.css",
      "require": "./dist/styles/*.css"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "styles"
  ],
  "keywords": [
    "min-kit",
    "components",
    "react",
    "taro",
    "miniapp",
    "miniprogram"
  ],
  "scripts": {
    "build": "unbuild",
    "test": "jest"
  },
  "dependencies": {
    "@min-kit/extends": "workspace:*",
    "@min-kit/hooks": "workspace:*",
    "@min-kit/shared": "workspace:*",
    "@min-kit/store": "workspace:*",
    "clsx": "^2.1.0"
  },
  "peerDependencies": {
    "@tarojs/components": ">=3.6",
    "@tarojs/plugin-platform-alipay": ">=3.6",
    "@tarojs/plugin-platform-h5": ">=3.6",
    "@tarojs/plugin-platform-jd": ">=3.6",
    "@tarojs/plugin-platform-qq": ">=3.6",
    "@tarojs/plugin-platform-swan": ">=3.6",
    "@tarojs/plugin-platform-tt": ">=3.6",
    "@tarojs/plugin-platform-weapp": ">=3.6",
    "@tarojs/react": ">=3.6",
    "@tarojs/taro": ">=3.6",
    "react": ">=17",
    "react-dom": ">=17"
  },
  "devDependencies": {
    "@min-kit/helper": "workspace:*",
    "@min-kit/jest": "workspace:^"
  }
}
