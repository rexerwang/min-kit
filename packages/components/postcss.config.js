module.exports = {
  plugins: {
    tailwindcss: {},
    'postcss-rem-to-responsive-pixel': {
      rootValue: 32,
      propList: ['*'],
      transformUnit: 'rpx',
    },
  },
}
