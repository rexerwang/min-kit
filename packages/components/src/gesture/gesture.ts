import type { ITouchEvent, TouchEventFunction } from '@tarojs/components'

export type SwipeDirection = 'left' | 'right' | 'up' | 'down'

export interface ISwipeEvent {
  horizontal?: boolean
  direction: SwipeDirection
  distance: number
}

export interface ITouchData {
  x: number
  y: number
  t: number
}

export function wrapTouch(handler: (data: ITouchData, evt: ITouchEvent) => void): TouchEventFunction {
  return (evt) => {
    const { touches, changedTouches } = evt
    const data = (touches?.length ? touches : changedTouches)?.at(0)
    if (data) {
      handler({ x: data.pageX, y: data.pageY, t: evt.timeStamp }, evt)
    }
  }
}

export function swipe(start: ITouchData, end: ITouchData) {
  const { x: x1, y: y1, t: t1 } = start
  const { x: x2, y: y2, t: t2 } = end

  const deltaX = Math.abs(x1 - x2)
  const deltaY = Math.abs(y1 - y2)

  if (deltaX > 80 || t2 - t1 < 250) {
    const direction: SwipeDirection = deltaX >= deltaY ? (x1 > x2 ? 'left' : 'right') : y1 > y2 ? 'up' : 'down'
    const horizontal = direction === 'left' || direction == 'right'
    const distance = horizontal ? deltaX : deltaY
    return { direction, horizontal, distance } satisfies ISwipeEvent
  }
}
