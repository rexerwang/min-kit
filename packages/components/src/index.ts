export * from './debug'
export * from './gesture'
export * from './icon'
export * from './img'
export * from './modal'
export * from './modal/drawer'
export * from './navigation-bar'
export * from './portal'
export * from './share'
export * from './text'
export type { ReactProps } from './types'
