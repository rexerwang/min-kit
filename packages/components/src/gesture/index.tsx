import { cloneElement, useRef } from 'react'

import { ISwipeEvent, ITouchData, swipe, wrapTouch } from './gesture'

type EventHandler<T> = (e: T) => void

export interface GestureProps {
  children: React.ReactElement
  onSwipe?: EventHandler<ISwipeEvent>
}

export function Gesture({ children, onSwipe }: GestureProps) {
  const startData = useRef<ITouchData>()

  const onTouchStart = wrapTouch((data) => {
    startData.current = data
  })

  const onTouchEnd = wrapTouch((data) => {
    if (startData.current && data) {
      const event = swipe(startData.current, data)
      event && onSwipe?.(event)
    }
  })

  return cloneElement(children, {
    onTouchStart,
    onTouchEnd,
  })
}

Gesture.displayName = 'MiniGesture'

export type { ISwipeEvent } from './gesture'
