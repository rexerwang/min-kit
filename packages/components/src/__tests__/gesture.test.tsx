import { act, createEvent, fireEvent, render, screen } from '@min-kit/jest'
import { View } from '@tarojs/components'

import { Gesture } from '../gesture'

describe('<Gesture />', () => {
  it('should render <Gesture /> toMatchSnapshot', () => {
    expect(
      render(
        <Gesture>
          <View>GestureTarget</View>
        </Gesture>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should determine swipe down', () => {
    const onSwipeStub = jest.fn()

    render(
      <Gesture onSwipe={onSwipeStub}>
        <View>GestureTarget</View>
      </Gesture>,
    )

    const el = screen.getByText(/GestureTarget/)

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchStart(el), { touches: [{ pageX: 0, pageY: 0 }] }))
    })

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchEnd(el), { touches: [{ pageX: 0, pageY: 10 }] }))
    })

    expect(onSwipeStub).toHaveBeenCalledWith({ horizontal: false, direction: 'down', distance: 10 })
  })

  it('should determine swipe up', () => {
    const onSwipeStub = jest.fn()

    render(
      <Gesture onSwipe={onSwipeStub}>
        <View>GestureTarget</View>
      </Gesture>,
    )

    const el = screen.getByText(/GestureTarget/)

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchStart(el), { touches: [{ pageX: 0, pageY: 10 }] }))
    })

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchEnd(el), { touches: [{ pageX: 0, pageY: 0 }] }))
    })

    expect(onSwipeStub).toHaveBeenCalledWith({ horizontal: false, direction: 'up', distance: 10 })
  })

  it('should determine swipe left', () => {
    const onSwipeStub = jest.fn()

    render(
      <Gesture onSwipe={onSwipeStub}>
        <View>GestureTarget</View>
      </Gesture>,
    )

    const el = screen.getByText(/GestureTarget/)

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchStart(el), { touches: [{ pageX: 10, pageY: 0 }] }))
    })

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchEnd(el), { touches: [{ pageX: 0, pageY: 0 }] }))
    })

    expect(onSwipeStub).toHaveBeenCalledWith({ horizontal: true, direction: 'left', distance: 10 })
  })

  it('should determine swipe right', () => {
    const onSwipeStub = jest.fn()

    render(
      <Gesture onSwipe={onSwipeStub}>
        <View>GestureTarget</View>
      </Gesture>,
    )

    const el = screen.getByText(/GestureTarget/)

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchStart(el), { changedTouches: [{ pageX: 0, pageY: 0 }] }))
    })

    act(() => {
      fireEvent(el, Object.assign(createEvent.touchEnd(el), { changedTouches: [{ pageX: 10, pageY: 0 }] }))
    })

    expect(onSwipeStub).toHaveBeenCalledWith({ horizontal: true, direction: 'right', distance: 10 })
  })
})
