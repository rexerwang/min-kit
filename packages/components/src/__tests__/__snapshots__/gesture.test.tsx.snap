// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gesture /> should render <Gesture /> toMatchSnapshot 1`] = `
<DocumentFragment>
  <taro-view-core
    class=""
  >
    <!---->
    GestureTarget
  </taro-view-core>
</DocumentFragment>
`;
